extensions/cms/src/cmsUtils.ts

```diff
@@ -186,19 +252,60 @@ export class CmsUtils {
 		});
 	}
 
+	public static getConnectionProfile(connection: azdata.connection.Connection): azdata.IConnectionProfile {
+		let connectionProfile: azdata.IConnectionProfile = {
+			connectionName: connection.options.connectionName,
+			serverName: connection.options.server,
+			databaseName: undefined,
+			userName: connection.options.user,
+			password: connection.options.password,
+			authenticationType: connection.options.authenticationType,
+			savePassword: connection.options.savePassword,
+			groupFullName: undefined,
+			groupId: undefined,
+			providerName: connection.providerName,
+			saveProfile: false,
+			id: connection.connectionId,
+			options: connection.options
+		};
+		return connectionProfile;
+	}
+
 	// Getters
 	public get registeredCmsServers(): ICmsResourceNodeInfo[] {
 		return this._registeredCmsServers;
 	}
 
-	public get connection(): Thenable<azdata.connection.Connection> {
-		return this.openConnectionDialog([cmsProvider], undefined, { saveConnection: false }).then((connection) => {
+	public connection(initialConnectionProfile?: azdata.IConnectionProfile): Thenable<azdata.connection.Connection> {
+		if (!initialConnectionProfile) {
+			initialConnectionProfile = {
+				connectionName: undefined,
+				serverName: undefined,
+				databaseName: undefined,
+				userName: undefined,
+				password: undefined,
+				authenticationType: undefined,
+				savePassword: undefined,
+				groupFullName: undefined,
+				groupId: undefined,
+				providerName: undefined,
+				saveProfile: undefined,
+				id: undefined,
+				options: {}
+			};
+		}
+		return this.openConnectionDialog([cmsProvider], initialConnectionProfile, { saveConnection: false }).then(async (connection) => {
 			if (connection) {
 				// remove group ID from connection if a user chose connection
 				// from the recent connections list
 				connection.options['groupId'] = null;
 				connection.providerName = mssqlProvider;
+				if (connection.options.savePassword) {
+					await this.savePassword(connection.options.user, connection.options.password);
+				}
 				return connection;
+			} else {
+				return Promise.reject();
 			}
```

Comment:
You should pretty much always return some sort of error message about why the error occured

Reply:
There's no error here, this is just the cancel button on connection dialog.

Reply:
There's nothing guaranteeing that to be the case. All you know is that openConnectionDialog returned with an undefined connection. But in this case it does unfortunately seem to be how we are choosing to expose that decision (personally it would be better if this returned a Result object that said what happened, so like DialogCancelled, ConnectionFailed, etc.).

But regardless - by using reject you're implying that this is an unexpected state and so should include a message about that state. Just something simple like 

throw new Error('User cancelled out of dialog');

although I don't know what benefits you see doing that over just returned undefined and having the caller check for that. 

---
In reply to: [292668264](https://github.com/microsoft/azuredatastudio/pull/5989#discussion_r292668264) [](ancestors = 292668264)